### PR TITLE
DEX-678 Expose relationships for given individual in GET API

### PR DESCRIPTION
**Pull request by karenc**
- Add validation for relationship_type_roles site setting

  Check that the structure of the relationship_type_roles follows
  something like:
  
  ```
  {
    "318cd874-4f52-45b3-9b6e-37fce007f40d": {
      "guid": "318cd874-4f52-45b3-9b6e-37fce007f40d",
      "label": "familial",
      "roles": [
        {
          "guid": "d68e45a9-ad60-44e1-9490-d8435e924c8c",
          "label": "mother"
        },
        {
          "guid": "64fd3382-47d4-42bb-8c88-df627981a49e",
          "label": "son"
        }
      ]
    },
  ```
  
  Update ExtraValidationSchema to handle nested many fields (like "roles").
  
  Update Notification*Schema so it no longer uses
  `<NotificationChannel.rest: 'restAPI'>` as the field name, instead use
  the value `"restAPI"` as the field name.

- Change relationship type and individual role to be guid

  We have this "relationship_type_roles" site setting which defines
  relationships that can be used for the site:
  
  ```
  {
      "49e85f81-c11a-42be-9097-d22c61345ed8": {
          "guid": "49e85f81-c11a-42be-9097-d22c61345ed8",
          "label": "familial",
          "roles": [
              {"guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c", "label": "mother"},
              {"guid": "ea5dbbb3-cc47-4d44-b460-2518a25dcb13", "label": "father"},
              {"guid": "4e38325a-46de-47e8-bb02-8ef0dbf54860", "label": "son"},
              {"guid": "8299f080-22ed-40da-8c06-af57ada64644", "label": "daughter"}
          ]
      }
  }
  ```
  
  Ben said we want to use the guid instead of strings for the relationship
  type and individual role.
  
  So instead of creating relationship like this:
  
  ```
  {
      "individual_1_guid": <guid>,
      "individual_2_guid": <guid>,
      "individual_1_role": "mother",
      "individual_2_role": "son",
      "type": "familial"
  }
  ```
  
  we are changing to this:
  
  ```
  {
      "individual_1_guid": <guid>,
      "individual_2_guid": <guid>,
      "individual_1_role_guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c",
      "individual_2_role_guid": "4e38325a-46de-47e8-bb02-8ef0dbf54860",
      "type_guid": "49e85f81-c11a-42be-9097-d22c61345ed8"
  }
  ```
  
  For GET relationship API, we would return the guids and labels:
  
  ```
  {
      ...
      "type_guid": "49e85f81-c11a-42be-9097-d22c61345ed8"
      "type_label": "familial",
      "individual_members": [
          {
              "individual_guid": <guid>,
              "individual_role_label": "mother",
              "individual_1_role_guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c",
          },
          {
              "individual_guid": <guid>,
              "individual_role_label": "son",
              "individual_2_role_guid": "4e38325a-46de-47e8-bb02-8ef0dbf54860",
          }
      ]
  }
  ```

- DEX-678 Expose relationships for given individual in GET API

  "relationships" is now included in the GET individuals API:
  

  ```
  GET /api/v1/individuals/<guid>
  
  {
    ...
    "relationships": [
      {
        "type_guid": "49e85f81-c11a-42be-9097-d22c61345ed8",
        "guid": "3b7c529c-0fbb-4e22-afbf-fccfe72ff394",
        "type_label": "Family",
        "individual_members": [
          {
            "individual_role_label": "Mother",
            "individual_guid": "fd1aa69c-5335-446e-9743-ba54298a8f74",
            "individual_role_guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c"
          },
          {
            "individual_role_label": "Calf",
            "individual_guid": "5a96ac3a-66b5-447c-be9b-6cf57c2f0777",
            "individual_role_guid": "ea5dbbb3-cc47-4d44-b460-2518a25dcb13"
          }
        ]
      }
    ],
    "guid": "2a4bd2ed-9622-49a8-ac62-f8ac8d7b119b",
    "updated": "2022-04-07T13:52:39.202409+00:00"
  }
  ```

### CODE REVIEW

**Comment by colinwkingen** [comment]
I think this will need a little more information, probably from a function. It looks like this shows the membership of this Individual and its role in the relationship, but it should also have the guid and role of the Individual it has a relationship with. 

Probably something like:
`
GET /api/v1/individuals/<guid>

{
  ...
  "relationships": [
    {
      "members": {
          [
              {
                        "individual_role": "Mother",
                        "individual_guid": "<guid>"
              },
              {
                        "individual_role": "Calf",
                        "individual_guid": "<guid>"
              }
          ]
      }
    }
  ],
  "guid": "2a4bd2ed-9622-49a8-ac62-f8ac8d7b119b",
  "updated": "2022-04-07T13:52:39.202409+00:00"
}
`

@brmscheiner curious if you have any notes on the shape of the response.

**Comment by brmscheiner** [comment]
@colinwkingen the major thing is that we added guids to relationship types and relationship roles. Here is what a relationship config looks like currently in site settings 
```
"value": {
                    "49e85f81-c11a-42be-9097-d22c61345ed8": {
                        "guid": "49e85f81-c11a-42be-9097-d22c61345ed8",
                        "label": "familial",
                        "roles": [
                            {
                                "guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c",
                                "label": "mother"
                            },
                            {
                                "guid": "ea5dbbb3-cc47-4d44-b460-2518a25dcb13",
                                "label": "father"
                            },
                            {
                                "guid": "4e38325a-46de-47e8-bb02-8ef0dbf54860",
                                "label": "son"
                            },
                            {
                                "guid": "8299f080-22ed-40da-8c06-af57ada64644",
                                "label": "daughter"
                            }
                        ]
                    }
                },
```
So on the individuals page we would need to be able to match these guids. Including `relationship_type_guid` as well as `relationship_role_guid` would do the trick 

```
  "relationships": [
    {
      "type_label": "familial",
      "type_guid": "<guid>",
      "members": {
          [
              {
                        "individual_role_label": "Mother",
                        "individual_role_guid": "<guid>",
                        "individual_guid": "<guid>"
              },
              {
                        "individual_role_label": "Calf",
                        "individual_role_guid": "<guid>",
                        "individual_guid": "<guid>"
              }
          ]
      }
    }
  ],
```

**Comment by karenc** [comment]
There's a `relationship_type_roles` site setting that looks like this on codex-qa:

```json
{                                          
  "318cd874-4f52-45b3-9b6e-37fce007f40d": {
    "guid": "318cd874-4f52-45b3-9b6e-37fce007f40d",
    "label": "a",
    "roles": [                             
      {                                    
        "guid": "d68e45a9-ad60-44e1-9490-d8435e924c8c",
        "label": "a1"
      },
      {
        "guid": "64fd3382-47d4-42bb-8c88-df627981a49e",
        "label": "a2"
      } 
    ] 
  },
  "c791b566-0a2c-4a5b-b5be-5a3cf643e812": {
    "guid": "c791b566-0a2c-4a5b-b5be-5a3cf643e812",
    "label": "b",
    "roles": [
      {
        "guid": "a1ed01d1-579a-4e8c-9350-f19672d9f465",
        "label": "b1"
      },
      {
        "guid": "d3850b6c-cf38-472f-88ba-62b26d1a0c14",
        "label": "b12"
      } 
    ] 
  } 
} 
```

For example, "a" would be "Family", "a1" would be "Mother" and "a2" would be "Calf".

The guids would be used to define the relationship types and roles.  The problem is our current implementation uses just strings:

https://github.com/WildMeOrg/houston/blob/3cbe84ef2e22bd1cc5f5056856934a092b5773ec/app/modules/relationships/models.py#L58

https://github.com/WildMeOrg/houston/blob/3cbe84ef2e22bd1cc5f5056856934a092b5773ec/app/modules/relationships/models.py#L32

Our tests use "Family", "Mother" and "Calf", not guids:

https://github.com/WildMeOrg/houston/blob/3cbe84ef2e22bd1cc5f5056856934a092b5773ec/tests/modules/relationships/resources/test_relationship_crud.py#L85-L93

So I think we need some changes to the APIs

**Comment by karenc** [comment]
> I'll add more tests to increase patch coverage :hourglass_flowing_sand:

The problem wasn't tests, I just needed to fix the coverage command when we check migrations.  We used `coverage run` instead of `coverage run --append` which meant the coverage from pytest was overwritten...  Fixed now

File: app/modules/site_settings/models.py [hunk 1]
```diff
@@ -228,13 +228,33 @@ def _get_default_value(cls, key):
         return def_val
 
     @classmethod
-    def set_key_value(cls, key, value):
-        if key == 'social_group_roles' and is_module_enabled('social_groups'):
+    def validate_social_group_roles(cls, value):
```
Comment (hwindsor) [inline app/modules/site_settings/models.py:231]: Neat idea, I like it.

**Comment by karenc** [comment]
@brmscheiner @Atticus29 This PR includes the following changes to the relationship and individuals APIs:

----

Instead of creating relationship like this:

```json
{
    "individual_1_guid": "fd1aa69c-5335-446e-9743-ba54298a8f74",
    "individual_2_guid": "5a96ac3a-66b5-447c-be9b-6cf57c2f0777",
    "individual_1_role": "mother",
    "individual_2_role": "son",
    "type": "familial"
}
```

we are changing to this:

```json
{
    "individual_1_guid": "fd1aa69c-5335-446e-9743-ba54298a8f74",
    "individual_2_guid": "5a96ac3a-66b5-447c-be9b-6cf57c2f0777",
    "individual_1_role_guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c",
    "individual_2_role_guid": "4e38325a-46de-47e8-bb02-8ef0dbf54860",
    "type_guid": "49e85f81-c11a-42be-9097-d22c61345ed8"
}
```

For GET relationship API, we would return the guids and labels:

```json
{
    ...
    "type_guid": "49e85f81-c11a-42be-9097-d22c61345ed8"
    "type_label": "familial",
    "individual_members": [
        {
            "individual_guid": "fd1aa69c-5335-446e-9743-ba54298a8f74",
            "individual_role_label": "mother",
            "individual_1_role_guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c",
        },
        {
            "individual_guid": "5a96ac3a-66b5-447c-be9b-6cf57c2f0777",
            "individual_role_label": "son",
            "individual_2_role_guid": "4e38325a-46de-47e8-bb02-8ef0dbf54860",
        }
    ]
}
```

----

"relationships" is now included in the GET individuals API:

```
GET /api/v1/individuals/<guid>
```

```json
{
  ...
  "relationships": [
    {
      "type_guid": "49e85f81-c11a-42be-9097-d22c61345ed8",
      "guid": "3b7c529c-0fbb-4e22-afbf-fccfe72ff394",
      "type_label": "Family",
      "individual_members": [
        {
          "individual_role_label": "Mother",
          "individual_guid": "fd1aa69c-5335-446e-9743-ba54298a8f74",
          "individual_role_guid": "1b62eb1a-0b80-4c2d-b914-923beba8863c"
        },
        {
          "individual_role_label": "Calf",
          "individual_guid": "5a96ac3a-66b5-447c-be9b-6cf57c2f0777",
          "individual_role_guid": "ea5dbbb3-cc47-4d44-b460-2518a25dcb13"
        }
      ]
    }
  ],
  "guid": "2a4bd2ed-9622-49a8-ac62-f8ac8d7b119b",
  "updated": "2022-04-07T13:52:39.202409+00:00"
}
```

**Review by colinwkingen — APPROVED** [review]
LGTM

**Comment by Atticus29** [comment]
> * NotificationChannel.rest

LGMT, too!

**Pull request merged**